d2ir/compile.go

```diff
@@ -167,6 +188,12 @@ func (c *compiler) resolveSubstitutions(varsStack []*Map, node Node) {
 						c.errorf(node.LastRef().AST(), `cannot substitute map variable "%s" as part of a string`, strings.Join(box.Substitution.IDA(), "."))
 						return
 					}
+					switch n := node.(type) {
```

Comment:
What's the reason for this?

Reply:
I'm not sure if we should do this. I wouldn't expect it at least.

Reply:
Only an explicit `null` should do this I would say.

Reply:
we check if primary is nil in places. https://github.com/alixander/d2/blob/ffdb5121f49e16e4f82f63cb034452881a77a91d/d2compiler/compile.go#L437

(if link is broken, line 1 of `compileReserved` function)

i think it makes sense. the primary was the substitution, which got resolved. and the resolution could be just a map with no primary.

Reply:
I don't understand what you wrote there. In regards to that check, why bother with that check? Shouldn't there be an error instead if you have a primary on `class`?

If I had a shape with a primary and a composite. Then I assigned it a composite variable, I wouldn't expect that the primary was overridden to nil even though the variable I assigned it had no primary.

Reply:
> Shouldn't there be an error instead if you have a primary on class?

no, class can be either primary or composite

> why bother with that check?

this check is everywhere in the compiler. `Primary() != nil` or `Primary() == nil`. i think that should continue to be a way of easily checking if the field has a primary or not.

> If I had a shape with a primary and a composite. Then I assigned it a composite variable, I wouldn't expect that the primary was overridden to nil even though the variable I assigned it had no primary.

I would. 
```d2
vars: { x: { a: b } }
a: 1
a: ${x}
```

it's definitely unexpected to me that it retains the 1 here

Reply:
Oh I see, I misunderstood the check in that code.

> this check is everywhere in the compiler. Primary() != nil or Primary() == nil. i think that should continue to be a way of easily checking if the field has a primary or not.

Why are you bringing this up? When did I assert that Primary != nil is not an adequate way to check if a field has a primary or not?

> it's definitely unexpected to me that it retains the 1 here

I'll defer.

Reply:
> When did I assert that Primary != nil is not an adequate way to check if a field has a primary or not?

you didn't. but if you substitute a composite with no primary and keep around the primary, then `Primary != nil` is no longer an adequate way to check. those places that check don't make the assumption that the primary could be a substitution that actually has no primary

Reply:
I'm super confused. Are you trying to use `Primary != nil` to check that there is no composite?